internal/provider/resource_cloud_service.go

```diff
@@ -8,6 +8,105 @@ import (
 	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
 )
 
+// used for all Hosted Clouds
+func resourceServicesHostedUpdate(ctx context.Context, d *schema.ResourceData, m interface{}, fn func(string, string) (*packetfabric.CloudServiceConnCreateResp, error)) diag.Diagnostics {
+	c := m.(*packetfabric.PFClient)
+	c.Ctx = ctx
+	var diags diag.Diagnostics
+	cloudCID, ok := d.GetOk("id")
+	if !ok {
+		return diag.Errorf("please provide a valid cloud service id")
+	}
+	if d.HasChange("aws_account_id") ||
+		d.HasChange("google_pairing_key") ||
+		d.HasChange("google_vlan_attachment_name") ||
+		d.HasChange("azure_service_key") ||
+		d.HasChange("ibm_account_id") ||
+		d.HasChange("ibm_bgp_asn") ||
+		d.HasChange("ibm_bgp_cer_cidr") ||
+		d.HasChange("ibm_bgp_ibm_cidr") ||
+		d.HasChange("vc_ocid") ||
+		d.HasChange("region") ||
+		d.HasChange("pop") ||
+		d.HasChange("zone") {
+		return diag.Errorf("only the description or speed field can be updated")
+	}
+	if desc, ok := d.GetOk("description"); !ok {
+		return diag.Errorf("please provide a valid description for Cloud Service")
+	} else {
+		resp, err := fn(desc.(string), cloudCID.(string))
+		if err != nil {
+			return diag.FromErr(err)
+		}
+		_ = d.Set("description", resp.Description)
+	}
+	// speed only for hosted cloud
+	if d.HasChange("speed") {
+		billing := packetfabric.BillingUpgrade{}
+		if speed, ok := d.GetOk("speed"); ok {
+			billing.Speed = speed.(string)
+		}
+		if _, err := c.ModifyBilling(cloudCID.(string), billing); err != nil {
+			return diag.FromErr(err)
+		}
+		if speed, ok := d.GetOk("speed"); ok {
+			_ = d.Set("speed", speed.(string))
+		}
+	}
+	return diags
+}
+
+// used for all Dedicated Clouds
+func resourceServicesDedicatedUpdate(ctx context.Context, d *schema.ResourceData, m interface{}, fn func(string, string) (*packetfabric.CloudServiceConnCreateResp, error)) diag.Diagnostics {
+	c := m.(*packetfabric.PFClient)
+	c.Ctx = ctx
+	var diags diag.Diagnostics
+	cloudCID, ok := d.GetOk("id")
+	if !ok {
+		return diag.Errorf("please provide a valid cloud service id")
+	}
+	if d.HasChange("description") ||
```

Comment:
Same thing here, I would suggest to create a validation function.

Reply:
Adding `checkUpdatableFieldsDedicatedCloud()` function